### ISSUE TITLE
Issue with a search count provider

### ISSUE THREAD

**Issue by mehalter** [rebelot/heirline.nvim#92]
I have a quick question regarding a seemingly simple provider that seems to error in a way that I don't quite understand. I am trying to set up a provider to show the search count mainly for when `cmdheight=0`. When I have this provider defined, if I go into visual mode and do a search/replace with `:'<,'>%s/\(` I get the following error:

```
E5108: Error executing lua Vim:E54: Unmatched \(
stack traceback:
        [C]: in function 'searchcount'
        [string "..."]: in function 'provider'
        ...k/packer/start/heirline.nvim/lua/heirline/statusline.lua:348: in function '_eval'
        ...k/packer/start/heirline.nvim/lua/heirline/statusline.lua:365: in function '_eval'
        ...k/packer/start/heirline.nvim/lua/heirline/statusline.lua:440: in function 'eval'
        ...te/pack/packer/start/heirline.nvim/lua/heirline/init.lua:71: in function <...te/pack/packer/start/heirline.nvim/lua/hei
rline/init.lua:66>
```

### Steps to Reproduce

0. Using the minimal `init.lua` file provided below
1. Run `nvim` and make sure to run `:PackerSync` and heirline is loaded
2. Go into visual mode, `v`
3. Go into command mode, `:` (which results in `:'<,'>`)
4. type `%s/\(`
5. as soon as the `(` is typed the error happens.

The error seems to be referring to the line `local search = vim.fn.searchcount()` but I'm not sure why it is causing heirline to crash in this instance

### Minimal `init.lua`

```lua
local install_path = vim.fn.stdpath("data") .. "/site/pack/packer/start/packer.nvim"

vim.g.loaded_remote_plugins = 1

if vim.fn.isdirectory(install_path) == 0 then
  vim.fn.system({ "git", "clone", "https://github.com/wbthomason/packer.nvim", install_path })
end

local packer = require("packer")
local use = packer.use

-- packer.reset()
packer.init({ git = { clone_timeout = -1 } })

use({ "wbthomason/packer.nvim" })

use({
  "rebelot/heirline.nvim",
  config = function()
    require("heirline").setup({
      {
        provider = function()
          local search = vim.fn.searchcount()
          if search.total then
            return string.format("%d/%d", search.current, math.min(search.total, search.maxcount))
          end
        end,
      },
    })
  end,
})
```

**Comment by rebelot** [comment]
I tested a bit and to me it seems that it has to do with the option `inccommand` being non empty. Because apparently `inccommand` updates the search pattern as you type and at the same time triggers a statusline redraw, it will cause `searchcount()` to be called at each keypress, which then panics on incomplete or wrong patterns. The error is the same as if you typed `:s/\(<CR>//`.

this is my version of it:
I used `pcall` to protect the call to search count which may fail during command preview, and to mitigate the `pcall` overhead I added a condition to show the component only when `hlsearch` highlights are visible (i.e., it will show up right after confirming a search and will hide after `:noh`)

```lua
local SearchCount = {
    condition = function()
        return vim.v.hlsearch ~= 0
    end,
    provider = function()
        local ok, search = pcall(vim.fn.searchcount)
        if ok and search.total then
            return string.format("%d/%d", search.current, math.min(search.total, search.maxcount))
        end
    end,
}
```
